### ISSUE TITLE
No -Wdangling-gsl warning if a gsl-pointer object is construct from a gsl-owner object in a container.

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#100384]
See the example below:

```c++
#include <vector>
#include <string>
#include <string_view>
using namespace std;

void test() {
  const std::string& a = std::vector<std::string>({""}).at(0); // good: give a warning
  string_view b = std::string(); // good: gives a warning
  string_view c = std::vector<std::string>({""}).at(0); // bad: no diagnostic emitted.
}
```

**Comment by pkasting** [comment]
May be related to https://github.com/llvm/llvm-project/issues/99685.

**Comment by usx95** [comment]
This looks different from https://github.com/llvm/llvm-project/issues/99685 which involves analysis beyond a single statement and definitely a form of reference tracking.

This one still looks approachable with the current infra of statement local analysis.

**Comment by usx95** [comment]
I think this falls in the same category of https://github.com/llvm/llvm-project/issues/100549

```cpp
```
https://godbolt.org/z/5oezd1ddn

**Comment by usx95** [comment]
Dumping non-vector related examples so that we do not miss them while fixing this
```cpp
#include <string>
#include <vector>

struct S {
    const std::string& get() const [[clang::lifetimebound]] { return s; }
    std::string s;
};
int main() { std::string_view c = S{"abcd"}.get(); }
```
https://godbolt.org/z/5oezd1ddn
Marking the conversion operator `lifetimebound` solves the problem here as well: https://godbolt.org/z/vfMaf17oq
https://godbolt.org/z/vfMaf17oq

**Comment by hokein** [comment]
In the original `vector.at` example, adding the lifetimebound annotation to the `.at` method would make the warning work, https://godbolt.org/z/rze5Yj8Ej.

It is unclear why, the interaction between lifetimebound and GSL is subtle, the triggered warning is `-Wdangling-gsl` (which is caused by the GSL-analysis code path), and the `at` should be handled [there](https://www.google.com/url?q=https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/CheckExprLifetime.cpp%23L300-L301&sa=D&source=docs&ust=1722941679226231&usg=AOvVaw1P_jrTS8m6-xef4D9nZKju).
